### ISSUE TITLE
Impossible to access shared USB devices after remount

### ISSUE THREAD

**Issue by Haythem-Zd** [cifsd-team/cifsd#100]
Hello,

I have an issue with cifsd server in a frequent use case: sharing a USB storage device mounted on the server machine.
My setup: The mount and the smb.conf update and the cifsd process restart are automated whenever a USB storage device is connected.

The way to reproduce this:

Connect USB => smb.conf updated and cifsd daemon is restarted
My smb.conf :
```
[global]
        netbios name = mytarget
        workgroup = lan
        server string = mytarget Sharing Files
        map to guest = bad user
[Haythem]
        comment = Share Haythem directory
        path = /mnt/sda1
        browsable = yes
        available = yes
        guest ok = yes
        guest only = yes
        writeable = yes
```

df command output:
> root@mytarget:~# df
> Filesystem           1K-blocks      Used Available Use% Mounted on
> ...
> /dev/sda1             63897564  14076672  49820892  22% /tmp/mnt/sda1

Up to this moment, everything is working. Great!

Now, after unplug and re-plug the USB device, sometimes, it is mounted with a different name.. say sdb1.
The user-space stuff works as expected and the paths in smb.conf and in df output command are updated and coherent (/mnt/sdb1 in this case and same share name). But the shares are no longer accessible by the clients (Windows and Linux GUI).

When trying to open a share from a Linux client (via GUI: **_smb://<server IP>_**), This error is printed on server machine:

> kcifsd: query_fs_info:4475: cannot create vfs path

And when activating kernel module debug logs:

> kcifsd: check_smb1_hdr:31: got SMB
> kcifsd: smb1_calc_size:195: SMB2 byte count 55, struct size : 8
> kcifsd: smb1_calc_size:198: SMB1 len 98
> kcifsd: smb_tree_connect_andx:431: tree connect request for tree HAYTHEM, dev_type : ?????
> kcifsd: cifsd_tcp_conn_handler_loop:209: RFC1002 header 72 bytes
> kcifsd: check_smb1_hdr:31: got SMB
> kcifsd: smb1_calc_size:195: SMB2 byte count 7, struct size : 30
> kcifsd: smb1_calc_size:198: SMB1 len 72
> kcifsd: query_fs_info:4475: cannot create vfs path
> kcifsd: smb_trans2:7176: smb_trans2 failed with error -2
> kcifsd: __process_request:205: Failed to process 50 [-2]
> kcifsd: cifsd_tcp_conn_handler_loop:209: RFC1002 header 76 bytes
> kcifsd: check_smb1_hdr:31: got SMB
> kcifsd: smb1_calc_size:195: SMB2 byte count 11, struct size : 30
> kcifsd: smb1_calc_size:198: SMB1 len 76
> kcifsd: smb_get_name:623: absoulte name = /mnt/sda1/
> kcifsd: query_path_info:3917: cannot get linux path for /mnt/sda1/, err -2
> kcifsd: smb_trans2:7176: smb_trans2 failed with error -2
> kcifsd: __process_request:205: Failed to process 50 [-2]
> 
When trying to open a shre from a Windows client (**_\\<server IP>\_**), This error is printed on server machine:

> kcifsd: cifsd_vfs_lookup_in_dir:1622: cannot open directory /mnt

And when activating kernel module debug logs:

> kcifsd: smb2_calc_size:302: SMB2 len 232
> kcifsd: smb2_open:2072: [smb2_open] 9
> kcifsd: smb2_open:2084: converted name = /mnt/sda1
> kcifsd: check_context_err:1977: find context ExtA err -2
> kcifsd: smb2_open:2216: get query maximal access context (timestamp : 1125968626319360)
> kcifsd: check_context_err:1977: find context TWrp err -2
> kcifsd: cifsd_vfs_lookup_in_dir:1622: cannot open directory /mnt
> kcifsd: cifsd_vfs_lookup_in_dir:1622: cannot open directory /mnt
> kcifsd: smb2_open:2253: can not get linux path for /mnt/sda1, rc = -22
> kcifsd: smb2_open:2364: returning as file does not exist

In both cases, at kernel module level, it looks like the server tries to use the old path **/mnt/sda1** and not the new (and valid) on **/mnt/sdb1**.

Is there a way to update this information in the server so it can keep up with the parameters in the conf file when changed (especially the share path)?

Thanks,
Haythem

**Comment by namjaejeon** [comment]
I have two questions.
 1. no problem when using samba ?
 2. Can you mannually umount /mnt/sda1 in problem situation(automount mounted /mnt/sdb1) ?

**Comment by sergey-senozhatsky** [comment]
Hi,

On (06/03/19 05:58), Haythem-Zd wrote:
>    I have an issue with cifsd server in a frequent use case: sharing a USB
>    storage device mounted on the server machine.
>    My setup: The mount and the smb.conf update and the cifsd process restart
>    are automated whenever a USB storage device is connected.

Hmm. kcifsd has a net-share configs cache; but we drop net-share config
from the cache once it has no references. Do you properly umount() USB
stick before you remount it?

Specifically we are interested in tree_disconnect request:

	umount /share
	// tree disconnect		- this should remove net-share from the cache
	// logoff
	detach USB
	attach USB
	update smb.conf
	restart cifsd
	mount /share
	// login
	// tree connect

	-ss

**Comment by Haythem-Zd** [comment]
@namjaejeon 
1. No, we do not have this issue with Samba. I just tested it.
2. I manually umounted /dev/sdb1 then remounted it with /dev/sda1 as mount point. And it is working again!

@sergey-senozhatsky 
We use the the mount/umount functions provided by the <sys/mount.h> API and it seems to be working well.

I should mention that the USB key I am testing with currently is NTFS formatted and I use the command ntfs-3g to mount it and the function umount() for the umount.

> root@mytarget:~# ntfs-3g -o nosuid,noexec,nodev,big_writes /dev/sda1 /mnt/sda1/

**Comment by Haythem-Zd** [comment]
Oh.. one more thing: right after umount(), the mount point directory (/mnt/sdxx) is deleted

**Comment by sergey-senozhatsky** [comment]
On (06/04/19 01:36), Haythem-Zd wrote:
>    [2]@sergey-senozhatsky
>    We use the the mount/umount functions provided by the <sys/mount.h> API
>    and it seems to be working well.
> 
>    I should mention that the USB key I am testing with currently is NTFS
>    formatted and I use the command ntfs-3g to mount it and the function
>    umount() for the umount.
> 
>      root@mytarget:~# ntfs-3g -o nosuid,noexec,nodev,big_writes /dev/sda1
>      /mnt/sda1/

I mean client mount/umount (e.g. mount -t cifs). IOW, before you
detach the USB stick, do all SMB client perform session logoff?

	-ss

**Comment by Haythem-Zd** [comment]
> before you detach the USB stick, do all SMB client perform session logoff?

No, it is very possible that the USB is disconnected from the sever machine even if there is some clients connected to some share(s).

> kcifsd has a net-share configs cache; but we drop net-share config from the cache once it has no references.

Does that mean if no tree disconnect or logoff is received, the cache will remain? In that case, I believe this is the cause as we can not guarantee clients disconnect before umounting the USB

**Comment by sergey-senozhatsky** [comment]
On (06/05/19 02:24), Haythem-Zd wrote:
>      kcifsd has a net-share configs cache; but we drop net-share config from
>      the cache once it has no references.
> 
>    Does that mean if no tree disconnect or logoff is received, the cache will
>    remain? In that case, I believe this is the cause as we can not guarantee
>    clients disconnect before umounting the USB

Hmm...
I believe there are bigger problems here. We can't kill all "abandoned"
sessions. They may hang in there with open files, which will prevent us
from destroying them.

	-ss

**Comment by Haythem-Zd** [comment]
is there a way to  "reset" a share in the cache from the user-space conf if the associated path is no longer accessible?

**Comment by namjaejeon** [comment]
Have you ever used deadtime parameter in smb.conf ? Let me know your issue is improved by using it.

**Comment by sergey-senozhatsky** [comment]
On (06/05/19 11:41), Namjae Jeon wrote:
> 
>    Have you ever used deadtime parameter in smb.conf ? Let me know your issue
>    is improved by using it.

cifsd_tcp_conn_alive() does

         if (atomic_read(&conn->stats.open_files_count) > 0)
                 return true;

which is always `true' for abandoned TCP connections. If client has
died, then it won't close its files and, thus, we will not terminate
the connection.

	-ss

**Comment by sergey-senozhatsky** [comment]
On (06/05/19 04:22), Haythem-Zd wrote:
>    is there a way to "reset" a share in the cache from the user-space conf if
>    the associated path is no longer accessible?

No, not yet. I'm not really sure if we need it (and how to handle
it). The real problem is that we keep dead TCP connections and sessions
around. Once we fix it net share cache invalidation should get back to
"sane".

	-ss

**Comment by sergey-senozhatsky** [comment]
On (06/05/19 23:59), Sergey Senozhatsky wrote:
> On (06/05/19 04:22), Haythem-Zd wrote:
> >    is there a way to "reset" a share in the cache from the user-space conf if
> >    the associated path is no longer accessible?
> 
> No, not yet. I'm not really sure if we need it (and how to handle
> it). The real problem is that we keep dead TCP connections and sessions
> around. Once we fix it net share cache invalidation should get back to
> "sane".

Well, that's not exactly true. Dead TCP connections is one scenario;
when net-share config is getting updated under us it's another thing.

Hot-reconfiguration is a difficult task and we don't handle it yet.


Any change you can change your setup to use hard/sym links instead
of devices names? E>g. /mnt/net-share where net-share either points
to sda1 or sdb1 or sdc1, etc.

	-ss

**Comment by Haythem-Zd** [comment]
I do not think that it solve the issue as we can have multiple USB devices / partitions shared at the same time.
We are now working on a solution that add a call from cifsd daemon to the kcifsd to close a share. The idea is inspired from Samba:
`#smbcontrol smbd close-share <ShareName>`
and do this right before umount.

**Comment by sergey-senozhatsky** [comment]
On (06/07/19 02:24), Haythem-Zd wrote:
>    I do not think that it solve the issue as we can have multiple USB devices
>    / partitions shared at the same time.
>    We are now working on a solution that add a call from cifsd daemon to the
>    kcifsd to close a share. The idea is inspired from Samba:
>    #smbcontrol smbd close-share <ShareName>
>    and do this right before umount.

I looked at it, for a bit. And dunno.
Share configs are shared between sessions/tcp connections. Killing
a net share for one sessions will kill it for all others, in a racy
manner. Adding a spinlock to struct net-share to protect ->path access
is just too much of an overkill.

	-ss

**Comment by namjaejeon** [comment]
@Haythem-Zd This issue is also coming on samba. I tested it with windows client. when connecting share through client, client send smb2_open abount usb mount point. So usb mount point is opened by smb server. and if we disconnect network drive, windows client send smb2 close about usb mount point, and smb server will close it but if you connect smb server using file explorer and close it, windows doesn't send smb2 close about mount point. so autumount script can not umount and resource busy because mount point is still being opened. and it make new device node(/dev/sdb1).
In conclusion, I think that it is not issue and same behavior with samba. Am I wrong ? Can you confirm if this issue doesn't happen on samba ?

Thanks!

**Comment by Haythem-Zd** [comment]
@namjaejeon 
You are right. I checked our storage service management scripts ant it turns out that we call 

> smbcontrol smbd close-share <ShareName>

 before umounting. That's why I did not reproduce the issue when I switched to Samba.

**Comment by namjaejeon** [comment]
> We are now working on a solution that add a call from cifsd daemon to the kcifsd to close a share. The idea is inspired from Samba:

great!, can you contribute the patch on mailing list if you fix it ?
Thanks!

**Comment by namjaejeon** [comment]
@Haythem-Zd ping ? Could you answer ?

**Comment by Haythem-Zd** [comment]
I close this issue since using the new tool cifsshareadd to delete shares is the perfect solution for my problem here.
Thank you!

**Comment by sergey-senozhatsky** [comment]
On (09/04/19 02:06), Haythem-Zd wrote:
>    I close this issue since using the new tool cifsshareadd to delete shares
>    is the perfect solution for my problem here.

We support only addition (dynamic cifsd reconfiguration)
of new shares at the moment. Anything else (-u or -d)
requires cifs-tools restart.

	-ss
